Created an indexer that can handle mpstat and proc-vmstat data.

mbruzek opened the pull request:
I added to`index-pbench` the individual handler that can index `proc-vmstat` and `mpstat` tool data. 

The `proc-vmstat` tool creates a finite set of csv files and a JSON document must be created for each column entry. As an example, here is the action object from one entry of the `proc-vmstat_compact_delta_sec.csv` file:  

<details><summary>JSON structure for one entry of the proc-vmstat_compact_delta_sec.csv file: </summary>
<p>

```json
    {
        "_index": "pbench.tool-data-proc-vmstat.2018-04-10",
        "_op_type": "create",
        "_source": {
            "@metadata": {
                "experiment": "pbench-user-benchmark_mbruzek-test-2_2018.04.10T19.01.19",
                "host": "b03-h01-1029p",
                "iteration": "1",
                "iterseqno": "1",
                "runid": "aec92de7ef92ede87062d974e00755a6",
                "runtstamp": "2018-04-10T19:01:19.927438829",
                "sample": "reference-result",
                "toolgroup": "default"
            },
            "@timestamp": "2018-04-10T19:01:24.054130",
            "proc-vmstat": {
                "vmstat": {
                    "compact": {
                        "migrate_scanned": {
                            "timestamp": "1523386884054.13",
                            "val": "0"
                        }
                    }
                }
            }
        }
```

</p>
</details>
<p></p>
The `mpstat` tool creates a variable set of csv files depending on the host machine CPU core count. Therefore we could not follow the pattern of naming the handler by the csv file basename. I created a new `@key_mappings` key in the tool handler dictionary with a regular expression, and handler key tuple. So if the csv basename matches the regular expression use the handler key to find the appropriate handler. This is extensible for tools in the future that may have different types of basenames map to different handler dictionaries. As an example here is the action object from one entry of the `cpu15_cpu15.csv` file:  

<details><summary>JSON structure for one entry of the cpu15_cpu15.csv file: </summary>
<p>

```json
    {
        "_index": "pbench.tool-data-mpstat.2018-04-10",
        "_op_type": "create",
        "_source": {
            "@metadata": {
                "experiment": "pbench-user-benchmark_mbruzek-test-2_2018.04.10T19.01.19",
                "host": "b03-h01-1029p",
                "iteration": "1",
                "iterseqno": "1",
                "runid": "73cd2a233e7491bbcb376ecd52b6978f",
                "runtstamp": "2018-04-10T19:01:19.927438829",
                "sample": "reference-result",
                "toolgroup": "default"
            },
            "@timestamp": "2018-04-10T19:01:24.000000",
            "mpstat": {
                "mpstat": {
                    "cpu": {
                        "sys": {
                            "timestamp": "1523386884000",
                            "val": "0.33"
                        }
                    },
                    "id": "cpu15"
                }
            }
        },
``` 

</p>
</details>
<p></p>

I am sure there is discussion to be had around the structure of each of these JSON structures. Please note I used the same `id` structure from the unified method:  
`[self.toolname][klass] = _dict_const([('id',identifier)])`

This PR also attempted to keep the paths similar, so both tools can be indexed by the same `_make_individual_source()` method:  
`datum[self.toolname][klass][metric][column] = { 'timestamp': row[0], 'val': val }`

Please review and give me feedback on what to change. Thanks!

mbruzek commented:
Hello @chaitanyaenr I took a look at this Travis failure. The unittests that I added in this PR (7.13, and 7.14) both pass. Each of the tar files for those tests only contain `proc-vmstat` csv files or `mpstat` csv files.

It looks like Travis failed on an older test `test-7.8` because the `state/test-7.8.tar.xz` file contains csv files for tools that were previously not indexed. I see mismatches on output [mpstat](https://travis-ci.org/distributed-system-analysis/pbench/jobs/382396272#L675) (functionality added in this PR) and [pidstat](https://travis-ci.org/distributed-system-analysis/pbench/jobs/382396272#L618) (functionality added in a different PR). Since the tools are now processed the output is different and that fails this Travis test.

mbruzek commented on server/pbench/bin/index-pbench:
```diff
@@ -618,6 +631,133 @@ _known_csv_handlers = {
+        '@prospectus': {
+            # For proc-vmstat .csv files, we want to individually index each
+            # column entry as its own JSON document.
+            'handling': 'individual'
```
I am new to pbench but my understanding on the `unify` method is it has to collect around a timestamp and identifier, and there was no ID in the proc-vmstat csv files (as there seems to be in iostat). 

After speaking with @ndokos and reading your other comments I believe I have come up with an alternate way to parse the output.

mbruzek commented on server/pbench/bin/index-pbench:
```diff
@@ -618,6 +631,133 @@ _known_csv_handlers = {
+            # column entry as its own JSON document.
+            'handling': 'individual'
+        },
+        'proc-vmstat_allocstall_delta_sec.csv': {
```
Yes I think it would be easier to parse the `proc-vmstat-stdout.txt` file for this information rather than the CSV files. I have a prototype that does this and creates a more flat data structure.

mbruzek commented on server/pbench/bin/index-pbench:
```diff
@@ -514,13 +514,26 @@ _known_csv_handlers = {
-            # For proc-interrupts .csv files, we want to individually index
-            # each column entry as its own JSON document
+            # For the mpstat .csv files, we want to individually index each
+            # column entry as its own JSON document.
```
Thanks for the comment here. I am new to pbench and the data structure I came up with was my best approximation looking at the other tools. I should be able to change the structure to something like this. Let me work on that.

mbruzek commented on server/pbench/bin/index-pbench:
```diff
@@ -514,13 +514,26 @@ _known_csv_handlers = {
+            # The mpstat tool produces csv files with names based on cpu cores.
+            # The amount of cpu cores can be different on each computer, so
+            # map a regular expression to a handler key for each file type.
+            (re.compile(r'(?P<id>cpu.+)_cpu(?P<cpuid>.+)\.csv'), 'generic_cpu'),
```
It might be _more_ complicated to use a regular expression as the key in this dictionary. Previously the `get_csv_files()` method used the file name as the key in the dict (`handler_rec = handler[fname]`). If the key can be a regular expression we have to iterate over all keys and attempt a match on the file name, when/if no matches are found then attempt to use the file name as key in the else condition. This is why I added an identifier at the top level.

There is no difference between the id number and `cpu`id. I thought it might come in handy to have the ID NN and the cpuNN for indexing but never used the second one.

portante replied:
Can we add something to the `@prospectus` entry to tell direct the code how to work with the handler entries?

It seems like it is worth the effort in the long run, since we'll need this for other tools going forward.

mbruzek replied:
I implemented a `handling` key and a `method` key in the prospectus as we discussed.

mbruzek commented:
I update this PR based on the review comments.

#### mpstat
Changes made to the structure of the data per the review comment:
A sample of the [mpstat data structure](http://pastebin.test.redhat.com/597808)

#### proc-vmstat
I changed this tool to use a new handler called `_known_output_handlers`, updated the code to be able to find the "toolname-stdout.txt" file and added the processing of output files as the last option behind json, and csv file handlers.

A sample of the [proc-vmstat data structure](http://pastebin.test.redhat.com/597815) 

I also fixed the other unit failures by removing the proc-vmstat and mpstat output from test-7.8.tar.xz.

Please take another look.

portante commented on server/pbench/bin/index-pbench:
```diff
@@ -838,7 +870,76 @@ class ToolData(object):
+            reader = csv['reader']
+            first_ts = None
+
+            while True:
```
The "unify" method has to read from all `.csv` files at once, so it needs to use the `next()` function on each `.csv` file generator, the `reader`.

Here we don't have to, because we are processing the entire `.csv` file before moving on to the next one.  So I think we can replace the
```
while True:
    try:
        row = next(reader)
    except StopIteration:
        # ...
        break
```

loop with:

```
for row in reader:
```

mbruzek replied:
I started from the `_make_source_unified(self)` method and didn't know we could use for. Good catch and I will change this.

portante commented on server/pbench/bin/index-pbench:
```diff
@@ -838,7 +870,76 @@ class ToolData(object):
+
+    def _make_source_stdout(self):
+        """
+        Read the stdout files and emit a record for each timestamp.
```
Can you enhance this doc text to read something like:

```
Read the given set of files one at a time, emitting a record
for each set of key/value pairs associated with a timestamp.

The format for files considered by this method are as follows:

 * each line of the file contains an ascii-numeric key and a
   numeric (integer or floating point) value separated by a ':'
 * the keyword, `timestamp`, is the recognized to be the
   timestamp value to be associated with all following
   key/value pairs until the next `timestamp` keyword
   is encountered
 * all key/value pairs are treated as fields of one JSON
   document with the associated timestamp to be indexed

An example file format:
    timestamp: 12345.00
    key0: 100.0
    key1: 100.1
    .
    .
    .
    keyN: 100.N
    timestamp: 12345.01
    key0: 100.0
    key1: 100.1
    .
    .
    .
    keyN: 100.N
    timestamp: ...
```

portante commented on server/pbench/bin/index-pbench:
```diff
@@ -863,15 +964,17 @@ class ToolData(object):
             gen = self._make_source_individual()
         elif self.handler['@prospectus']['handling'] == 'json':
             gen = self._make_source_json()
+        elif self.handler['@prospectus']['handling'] == 'stdout':
```
Can we consider renaming `stdout` to something closer to the contents we are targeting?  Perhaps, though kind of long: `periodic_timestamp_key_value`?

portante commented on server/pbench/bin/index-pbench:
```diff
@@ -676,24 +702,30 @@ class ToolData(object):
 
         self.csv_files = None
         self.json_files = None
+        self.stdout_files = None
```
Would you be willing to perform some additional refactoring to smooth this out a bit?

Can we merge all the json, csv, and output, handlers into one structure, `_known_handlers`, indexed by tool?  Then have one `staticmethod` called, `get_files` which picks the correct `get_*_files` method based on the handler in play?

mbruzek replied:
This kind of simplification forces me to change code that this PR did not already alter, such as the unified csv handling and the json handling. I don't want to break anything, but will comply with the request.

mbruzek commented:
> It seems bad to change another unit test to remove data from it. Can we update that unit test to work with the new indexing instead?

When the test-7.8 was written index-pbench didn't produce any output for mpstat or proc-vmstat, but with this PR the output now contains some mpstat data that does not match the expected gold result and fails the unittest 7.8 that has nothing to do with mpstat, or proc-vmstat. The test-7.8 does not exercise any mpstat or proc-vmstat content so I removed those outputs from the tar file (making it smaller and more specific to test-7.8).

portante commented:
@mbruzek, can you restore test-7.8 to what it was and ensure that the newly generated data sets are correct? Thanks!

mbruzek commented:
@portante I updated this PR with the original test-7.8.tar.xz and changed the gold/test-7.8_output.txt as you requested. 

I also implemented the more refined way to get files that we talked about. 

All unittests pass on my machine and I hope travis agrees.

mbruzek commented:
@portante This is passing all tests and is the product of what we discussed. Do you have any other feedback on this PR?

portante commented:
@mbruzek, can you rebase this work on top of master and update?  We'd like to get this in for 0.53 of the pbench-agent/server release.

mbruzek commented:
@portante yes of course. A lot has changed since my last commit so it might take me a while to get everything sorted, but I will start working on it.

portante commented on server/pbench/bin/index-pbench:
```diff
@@ -881,7 +915,109 @@ class ToolData(object):
+                for line in file_object:
+                    ts = None
+                    if line.startswith('timestamp:'):
+                        # timestamp delimits records, yeild last record.
```
s/yeild/yield

portante commented on server/pbench/bin/index-pbench:
```diff
@@ -881,7 +915,109 @@ class ToolData(object):
+            with open(path, 'r') as file_object:
+                record = None
+                for line in file_object:
+                    ts = None
```
Is this initialization necessary?

mbruzek replied:
Probably not.

portante commented on server/pbench/bin/index-pbench:
```diff
@@ -542,9 +542,43 @@ _known_handlers = {
+                # The mpstat tool produces csv files with names based on cpu cores.
+                # The amount of cpu cores can be different on each computer, so
+                # map a regular expression to a handler key for each file type.
+                'pattern': re.compile(r'(?P<id>cpu.+)_cpu(?P<cpuid>.+)\.csv'),
```
Where are we using the `cpuid` value?

mbruzek replied:
Not using cpuid so I will take it out.

portante commented on server/pbench/bin/index-pbench:
```diff
@@ -1011,6 +1149,29 @@ class ToolData(object):
+                    handler_rec = rec
+                    break
+
+            assert handler_rec is not None
```
Instead of an `assert` it seems like you can use the `for/else` construct here to where the `else` clause is executed when the for loop goes through all its iterations [1].

[1] https://docs.python.org/2/reference/compound_stmts.html#the-for-statement

mbruzek replied:
Yeah I copied that from `get_csv_files()` and thought it was a good idea.

portante commented on server/pbench/bin/index-pbench:
```diff
@@ -881,7 +915,109 @@ class ToolData(object):
+
+                source_id = _make_source_id(datum)
+                yield datum, source_id
+                #yield datum
```
Seems like this line can be removed ...

mbruzek replied:
yep

portante commented on server/pbench/bin/index-pbench:
```diff
@@ -881,7 +915,109 @@ class ToolData(object):
+                    if idx == 0:
+                        time_val = float(val)/1000
+                        utc = datetime.utcfromtimestamp(time_val)
+                        # Convert the first timestamp to a formatted string.
```
Is this converting only the first timestamp found or all timestamps found?  Seems like the `first_ts` is really just the "current" timestamp.

mbruzek replied:
correct

portante commented on server/pbench/bin/index-pbench:
```diff
@@ -881,7 +915,109 @@ class ToolData(object):
+                        try:
+                            # This handler has the id in the filename pattern.
+                            datum[klass][metric] = _dict_const([('id', match.group('id'))])
+                        except IndexError as ie:
```
Not sure you need `as ie` here.

mbruzek replied:
removed

portante commented on server/pbench/bin/index-pbench:
```diff
@@ -881,7 +915,109 @@ class ToolData(object):
+
+         * each line of the file contains an ascii-numeric key and a
+           numeric (integer or floating point) value separated by a ':'
+         * the keyword, `timestamp`, is the recognized to be the
```
s/is the/is/

mbruzek replied:
good catch.

portante commented on server/pbench/bin/index-pbench:
```diff
@@ -881,7 +915,109 @@ class ToolData(object):
+            basename = output_file['basename']
+            handler = output_file['handler']
+            klass = handler['class']
+            # file_object = output_file['file_object']
```
Do we need this comment?

mbruzek replied:
no

mbruzek commented:
Rebased **again** with all your comments. @portante please take a look.
